File: typecast-kotlin/src/main/kotlin/com/neosapience/models/CustomVoice.kt
package com.neosapience.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Response of POST /v1/custom-voices/instant-clone — custom voice (created via instant cloning) metadata.
 *
 * The [voiceId] field has the `"uc_"` prefix and can be used directly as
 * `voice_id` in [com.neosapience.TypecastClient.textToSpeech] calls.
 */
@Serializable
data class CustomVoice(
    @SerialName("voice_id") val voiceId: String,
    @SerialName("name") val name: String,
    @SerialName("model") val model: String,
    val source: String? = null,
    val status: String? = null,
    val error: String? = null,
    @SerialName("created_at") val createdAt: String? = null,
) {
    companion object {
        /** Maximum audio file size accepted by cloneVoice (25 MB). Matches typecast-api. */
        const val CLONING_MAX_FILE_SIZE: Long = 25L * 1024 * 1024

        /** Minimum allowed length for the voice name passed to cloneVoice. */
        const val NAME_MIN_LENGTH: Int = 1

        /** Maximum allowed length for the voice name passed to cloneVoice. */
        const val NAME_MAX_LENGTH: Int = 30
    }
}

/** A named audio sample for professional voice cloning. */
data class CustomVoiceFile(val filename: String, val data: ByteArray)

File: typecast-php/src/Models/CustomVoice.php
<?php

declare(strict_types=1);

namespace Neosapience\Typecast\Models;

/**
 * Response of POST /v1/custom-voices/instant-clone — custom voice (created via instant cloning) metadata.
 *
 * The voiceId has the "uc_" prefix and can be used directly as the voiceId
 * parameter in textToSpeech calls.
 */
final class CustomVoice
{
    /** Maximum audio file size accepted by cloneVoice (25 MB). */
    public const CLONING_MAX_FILE_SIZE = 25 * 1024 * 1024;

    /** Minimum allowed length for a custom voice name. */
    public const NAME_MIN_LENGTH = 1;

    /** Maximum allowed length for a custom voice name. */
    public const NAME_MAX_LENGTH = 30;

    public function __construct(
        public readonly string $voiceId,
        public readonly string $name,
        public readonly string $model,
        public readonly ?string $source = null,
        public readonly ?string $status = null,
        public readonly ?string $error = null,
        public readonly ?string $createdAt = null,
    ) {}

    /**
     * Create from an API JSON response array.
     *
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        return new self(
            voiceId: (string) ($data['voice_id'] ?? ''),
            name:    (string) ($data['name']     ?? ''),
            model:   (string) ($data['model']    ?? ''),
            source:  isset($data['source']) ? (string) $data['source'] : null,
            status:  isset($data['status']) ? (string) $data['status'] : null,
            error:   isset($data['error']) ? (string) $data['error'] : null,
            createdAt: isset($data['created_at']) ? (string) $data['created_at'] : null,
        );
    }
}
